mcweb/backend/sources/tasks.py

```diff
@@ -315,3 +316,95 @@ def schedule_scrape_source(source_id, user):
                           remove_existing_tasks=True)
     return return_task(task)
 
+
+DAYS_BACK = 365  # Number of days to look back for story analysis
+MIN_STORY_COUNT = 100  # Minimum number of stories required for a valid source
+START_DATE = dt.datetime(2000, 1, 1) # Possible earliest source publication date
+END_DATE = dt.datetime.now()
+
+def analyze_sources(batch_size: int, analysis_type: str, start_date: dt.datetime) -> List[Dict[str, str]]:
+    """
+    Generalized function to analyze sources.
+    Args:
+        batch_size (int): Number of sources to process in each batch.
+        analysis_type (str): Type of analysis to perform ("language" or "publication_date").
+    Returns:
+        List[Dict[str, str]]: A list of dictionaries containing source IDs and their analyzed data.
+    """
+    # TO DO:xavier About 1693 sources have name__isnull=True but have homepage
+    # getting canonical domain from urls.canonical_domain(homepage) makes a HTTP request, NOT IDEAL !!!
+    sources = Source.objects.filter(name__isnull=False)
+    total_sources = sources.count()
+    logger.info(f"Starting {analysis_type} analysis for {total_sources} sources.")
+
+    updated_sources = []
+
+    for i in range(0, total_sources, batch_size):
+        batch = sources[i:i + batch_size]
+        for source in batch:
+            try:
+                pq = ParsedQuery(
+                    start_date=start_date,
+                    end_date=END_DATE,
+                    query_str=f"canonical_domain:{source.name}",
+                    provider_props={},
+                    provider_name="onlinenews-mediacloud",
+                    api_key=None,
+                    base_url=None
+                )
+                provider = pq_provider(pq)
+                results = provider._overview_query(pq.query_str, start_date, END_DATE)
```

Comment:
If this is a network call, we may need to add a small delay/jitter since we're in a loop.

Reply:
Requests via the web are ratelimited at 100/m - we should use the same value here.